NPE when opening the options dialog

Using current release/5.0 branch, if I click More -> Options, then I get this error:

```
ERROR 22:55:55 DCUncaughtExceptionHandler - Thread AWT-EventQueue-0 threw uncaught exception
com.google.inject.ProvisionException: Guice provision errors:

1) Error injecting constructor, java.lang.NullPointerException
  at org.datacleaner.windows.OptionsDialog.<init>(OptionsDialog.java:94)
  while locating org.datacleaner.windows.OptionsDialog

1 error
        at com.google.inject.internal.InjectorImpl$4.get(InjectorImpl.java:987)
        at org.datacleaner.windows.AnalysisJobBuilderWindowImpl$6.actionPerformed(AnalysisJobBuilderWindowImpl.java:912)
        at javax.swing.AbstractButton.fireActionPerformed(Unknown Source)
        at javax.swing.AbstractButton$Handler.actionPerformed(Unknown Source)
        at javax.swing.DefaultButtonModel.fireActionPerformed(Unknown Source)
        at javax.swing.DefaultButtonModel.setPressed(Unknown Source)
        at javax.swing.AbstractButton.doClick(Unknown Source)
        at javax.swing.plaf.basic.BasicMenuItemUI.doClick(Unknown Source)
        at javax.swing.plaf.basic.BasicMenuItemUI$Handler.mouseReleased(Unknown Source)
        at java.awt.Component.processMouseEvent(Unknown Source)
        at javax.swing.JComponent.processMouseEvent(Unknown Source)
        at java.awt.Component.processEvent(Unknown Source)
        at java.awt.Container.processEvent(Unknown Source)
        at java.awt.Component.dispatchEventImpl(Unknown Source)
        at java.awt.Container.dispatchEventImpl(Unknown Source)
        at java.awt.Component.dispatchEvent(Unknown Source)
        at java.awt.LightweightDispatcher.retargetMouseEvent(Unknown Source)
        at java.awt.LightweightDispatcher.processMouseEvent(Unknown Source)
        at java.awt.LightweightDispatcher.dispatchEvent(Unknown Source)
        at java.awt.Container.dispatchEventImpl(Unknown Source)
        at java.awt.Window.dispatchEventImpl(Unknown Source)
        at java.awt.Component.dispatchEvent(Unknown Source)
        at java.awt.EventQueue.dispatchEventImpl(Unknown Source)
        at java.awt.EventQueue.access$400(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.awt.EventQueue$3.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$1.doIntersectionPrivilege(Unknown Source)
        at java.security.ProtectionDomain$1.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue$4.run(Unknown Source)
        at java.awt.EventQueue$4.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$1.doIntersectionPrivilege(Unknown Source)
        at java.awt.EventQueue.dispatchEvent(Unknown Source)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForFilter(Unknown Source)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.pumpEvents(Unknown Source)
        at java.awt.EventDispatchThread.run(Unknown Source)
Caused by: java.lang.NullPointerException
        at org.datacleaner.windows.OptionsDialog$RemoteComponentsTab.getRemoteComponentsServer(OptionsDialog.java:602)
        at org.datacleaner.windows.OptionsDialog$RemoteComponentsTab.<init>(OptionsDialog.java:501)
        at org.datacleaner.windows.OptionsDialog.<init>(OptionsDialog.java:107)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.lang.reflect.Constructor.newInstance(Unknown Source)
        at com.google.inject.internal.DefaultConstructionProxyFactory$2.newInstance(DefaultConstructionProxyFactory.java:85)
        at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:85)
        at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:254)
        at com.google.inject.internal.InjectorImpl$4$1.call(InjectorImpl.java:978)
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1024)
        at com.google.inject.internal.InjectorImpl$4.get(InjectorImpl.java:974)
        ... 39 more
```

--- a/api/src/main/java/org/datacleaner/configuration/RemoteServerConfiguration.java
+++ b/api/src/main/java/org/datacleaner/configuration/RemoteServerConfiguration.java
@@ -25,6 +25,7 @@ import java.util.List;
  * Container for all remote servers' (DataCleaner Monitors) configuration.
  */
 public interface RemoteServerConfiguration {
+    
     /**
      * Getter for a flag indicating whether all components coming from various remote servers are shown together
      * (in the component's tree, components' context menu).

--- a/components/basic-transformers/src/test/java/org/datacleaner/beans/transform/RemoveSubstringTransformerTest.java
+++ b/components/basic-transformers/src/test/java/org/datacleaner/beans/transform/RemoveSubstringTransformerTest.java
@@ -44,7 +44,7 @@ public class RemoveSubstringTransformerTest {
         // Just to test that other types doesn't sent it spinning
         final InputColumn<?> booleanColumn = new MockInputColumn<>("booleanCol", Boolean.class);
 
-        final InputColumn[] subtractColumns =
+        final InputColumn<?>[] subtractColumns =
                 new InputColumn[] { stringColumn, numberColumn, listColumn, booleanColumn };
 
         _inputRow = new MockInputRow().put(baseColumn, "hello goodbye, 5, a, 2, c, true, false")

--- a/components/el/pom.xml
+++ b/components/el/pom.xml
@@ -21,7 +21,6 @@
 		<dependency>
 			<groupId>javax.el</groupId>
 			<artifactId>javax.el-api</artifactId>
-			<version>2.2.4</version>
 		</dependency>
 	</dependencies>
 </project>

--- a/desktop/api/src/main/java/org/datacleaner/util/IconUtils.java
+++ b/desktop/api/src/main/java/org/datacleaner/util/IconUtils.java
@@ -216,7 +216,7 @@ public final class IconUtils {
         boolean serverDown = false;
 
         if (descriptor instanceof RemoteTransformerDescriptor) {
-            if (!((RemoteTransformerDescriptor) descriptor).getRemoteDescriptorProvider().isServerUp()) {
+            if (!((RemoteTransformerDescriptor<?>) descriptor).getRemoteDescriptorProvider().isServerUp()) {
                 serverDown = true;
             }
         }

--- a/desktop/ui/src/main/java/org/datacleaner/windows/OptionsDialog.java
+++ b/desktop/ui/src/main/java/org/datacleaner/windows/OptionsDialog.java
@@ -119,12 +119,10 @@ public class OptionsDialog extends AbstractWindow {
         _tabbedPane.addTab("Extensions", imageManager.getImageIcon(IconUtils.PLUGIN, IconUtils.ICON_SIZE_TAB),
                 extensionPackagesPanel);
 
-        _tabbedPane.setUnclosableTab(0);
-        _tabbedPane.setUnclosableTab(1);
-        _tabbedPane.setUnclosableTab(2);
-        _tabbedPane.setUnclosableTab(3);
-        _tabbedPane.setUnclosableTab(4);
-        _tabbedPane.setUnclosableTab(5);
+        final int tabCount = _tabbedPane.getTabCount();
+        for (int i = 0; i < tabCount; i++) {
+            _tabbedPane.setUnclosableTab(i);
+        }
     }
 
     private URL getDataCleanerConfigurationFileURI() {

--- a/desktop/ui/src/main/java/org/datacleaner/windows/OptionsDialog.java
+++ b/desktop/ui/src/main/java/org/datacleaner/windows/OptionsDialog.java
@@ -19,33 +19,38 @@
  */
 package org.datacleaner.windows;
 
-import java.awt.*;
+import java.awt.BorderLayout;
+import java.awt.Component;
+import java.awt.Dimension;
+import java.awt.GridBagConstraints;
+import java.awt.Image;
 import java.awt.event.ActionEvent;
 import java.awt.event.ActionListener;
 import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.text.NumberFormat;
 import java.util.HashMap;
-import java.util.List;
 import java.util.Map;
 
 import javax.inject.Inject;
-import javax.swing.*;
+import javax.swing.JButton;
+import javax.swing.JCheckBox;
+import javax.swing.JComponent;
+import javax.swing.JFileChooser;
+import javax.swing.JLabel;
+import javax.swing.JPanel;
+import javax.swing.JPasswordField;
+import javax.swing.JTextField;
+import javax.swing.Timer;
 import javax.swing.border.EmptyBorder;
 import javax.swing.event.DocumentEvent;
 
 import org.datacleaner.bootstrap.WindowContext;
 import org.datacleaner.configuration.DataCleanerConfiguration;
 import org.datacleaner.configuration.DataCleanerConfigurationUpdater;
-import org.datacleaner.configuration.JaxbConfigurationReader;
 import org.datacleaner.configuration.RemoteServerConfiguration;
 import org.datacleaner.configuration.RemoteServerData;
-import org.datacleaner.configuration.jaxb.Configuration;
-import org.datacleaner.configuration.jaxb.RemoteComponentServerType;
-import org.datacleaner.configuration.jaxb.RemoteComponentsType;
 import org.datacleaner.job.concurrent.MultiThreadedTaskRunner;
 import org.datacleaner.job.concurrent.TaskRunner;
 import org.datacleaner.panels.DCBannerPanel;
@@ -89,8 +94,8 @@ public class OptionsDialog extends AbstractWindow {
 
     @Inject
     protected OptionsDialog(WindowContext windowContext, DataCleanerConfiguration configuration,
-                            UserPreferences userPreferences, DatabaseDriversPanel databaseDriversPanel,
-                            ExtensionPackagesPanel extensionPackagesPanel) {
+            UserPreferences userPreferences, DatabaseDriversPanel databaseDriversPanel,
+            ExtensionPackagesPanel extensionPackagesPanel) {
         super(windowContext);
         _userPreferences = userPreferences;
         _configuration = configuration;
@@ -104,8 +109,9 @@ public class OptionsDialog extends AbstractWindow {
                 databaseDriversPanel);
         _tabbedPane.addTab("Network", imageManager.getImageIcon("images/menu/network.png", IconUtils.ICON_SIZE_TAB),
                 getNetworkTab());
-        _tabbedPane.addTab("Remote components", imageManager.getImageIcon("images/menu/remote-components.png",
-                IconUtils.ICON_SIZE_TAB), new RemoteComponentsTab());
+        _tabbedPane.addTab("Remote components",
+                imageManager.getImageIcon("images/menu/remote-components.png", IconUtils.ICON_SIZE_TAB),
+                new RemoteComponentsTab());
         _tabbedPane.addTab("Performance",
                 imageManager.getImageIcon("images/menu/performance.png", IconUtils.ICON_SIZE_TAB), getPerformanceTab());
         _tabbedPane.addTab("Memory", imageManager.getImageIcon("images/menu/memory.png", IconUtils.ICON_SIZE_TAB),
@@ -123,8 +129,8 @@ public class OptionsDialog extends AbstractWindow {
 
     private URL getDataCleanerConfigurationFileURI() {
         try {
-            String path = _configuration.getHomeFolder().toFile().getAbsolutePath() +
-                    File.separator + CONFIGURATION_FILE_NAME;
+            String path = _configuration.getHomeFolder().toFile().getAbsolutePath() + File.separator
+                    + CONFIGURATION_FILE_NAME;
             File file = new File(path);
 
             return file.toURI().toURL();
@@ -318,11 +324,10 @@ public class OptionsDialog extends AbstractWindow {
         TaskRunner taskRunner = _configuration.getEnvironment().getTaskRunner();
         WidgetUtils.addToGridBag(new JLabel("Task runner type:"), panel, 0, row);
         WidgetUtils.addToGridBag(new JLabel(taskRunner.getClass().getSimpleName()), panel, 1, row);
-        WidgetUtils
-                .addToGridBag(
-                        new HelpIcon(
-                                "The task runner is used to determine the execution strategy of Analysis jobs. The most common strategy for this is to use a multithreaded task runner which will spawn several threads to enable concurrent execution of jobs."),
-                        panel, 2, row);
+        WidgetUtils.addToGridBag(
+                new HelpIcon(
+                        "The task runner is used to determine the execution strategy of Analysis jobs. The most common strategy for this is to use a multithreaded task runner which will spawn several threads to enable concurrent execution of jobs."),
+                panel, 2, row);
 
         if (taskRunner instanceof MultiThreadedTaskRunner) {
             int numThreads = ((MultiThreadedTaskRunner) taskRunner).getNumThreads();
@@ -338,15 +343,14 @@ public class OptionsDialog extends AbstractWindow {
         StorageProvider storageProvider = _configuration.getEnvironment().getStorageProvider();
         WidgetUtils.addToGridBag(new JLabel("Storage provider type:"), panel, 0, row);
         WidgetUtils.addToGridBag(new JLabel(storageProvider.getClass().getSimpleName()), panel, 1, row);
-        WidgetUtils
-                .addToGridBag(
-                        new HelpIcon(
-                                "The storage provider is used for staging data during and after analysis, typically to store the results on disk in stead of holding everything in memory."),
-                        panel, 2, row);
+        WidgetUtils.addToGridBag(
+                new HelpIcon(
+                        "The storage provider is used for staging data during and after analysis, typically to store the results on disk in stead of holding everything in memory."),
+                panel, 2, row);
 
         row++;
-        DCLabel descriptionLabel = DCLabel
-                .darkMultiLine("Performance options are currently not configurable while you're running the application. "
+        DCLabel descriptionLabel = DCLabel.darkMultiLine(
+                "Performance options are currently not configurable while you're running the application. "
                         + "You need to edit the applications configuration file for this. The configuration file is named "
                         + "<b>" + CONFIGURATION_FILE_NAME + "</b> and is located in the root of the folder where "
                         + "you've installed DataCleaner.");
@@ -487,7 +491,6 @@ public class OptionsDialog extends AbstractWindow {
         return _userPreferences;
     }
 
-
     private class RemoteComponentsTab extends DCPanel {
         private static final long serialVersionUID = 1L;
         private int left = 0;
@@ -498,7 +501,6 @@ public class OptionsDialog extends AbstractWindow {
         private int maxWeightx = 100;
         private int weightx = 1;
         private int weighty = 0;
-        private RemoteComponentServerType remoteServer = getRemoteComponentsServer();
         private JTextField usernameTextField;
         private JPasswordField passwordTextField;
 
@@ -535,12 +537,16 @@ public class OptionsDialog extends AbstractWindow {
         }
 
         private void setupFields() {
-            if (remoteServer == null) {
+            final RemoteServerConfiguration remoteServerConfiguration = _configuration.getEnvironment()
+                    .getRemoteServerConfiguration();
+            if (remoteServerConfiguration == null || remoteServerConfiguration.getServerList().isEmpty()) {
                 usernameTextField.setEnabled(false);
                 passwordTextField.setEnabled(false);
             } else {
-                setupFieldForRemoteComponentsTab(usernameTextField, remoteServer.getUsername());
-                setupFieldForRemoteComponentsTab(passwordTextField, remoteServer.getPassword());
+                final RemoteServerData remoteServerData = remoteServerConfiguration.getServerList().get(0);
+
+                setupFieldForRemoteComponentsTab(usernameTextField, remoteServerData.getUsername());
+                setupFieldForRemoteComponentsTab(passwordTextField, remoteServerData.getPassword());
             }
         }
 
@@ -554,12 +560,12 @@ public class OptionsDialog extends AbstractWindow {
 
             JPanel spaceFiller = new JPanel();
             spaceFiller.setBackground(WidgetUtils.COLOR_DEFAULT_BACKGROUND);
-            WidgetUtils.addToGridBag(new JLabel(labelText), this, left, row, 1, 1,
-                    GridBagConstraints.LINE_START, padding, weightx, weighty, GridBagConstraints.HORIZONTAL);
-            WidgetUtils.addToGridBag(field, this, left + 1, row, 1, 1,
-                    GridBagConstraints.LINE_START, padding, weightx, weighty, GridBagConstraints.HORIZONTAL);
-            WidgetUtils.addToGridBag(spaceFiller, this, left + 2, row, 1, 1,
-                    GridBagConstraints.LINE_END, padding, maxWeightx, weighty, GridBagConstraints.REMAINDER);
+            WidgetUtils.addToGridBag(new JLabel(labelText), this, left, row, 1, 1, GridBagConstraints.LINE_START,
+                    padding, weightx, weighty, GridBagConstraints.HORIZONTAL);
+            WidgetUtils.addToGridBag(field, this, left + 1, row, 1, 1, GridBagConstraints.LINE_START, padding, weightx,
+                    weighty, GridBagConstraints.HORIZONTAL);
+            WidgetUtils.addToGridBag(spaceFiller, this, left + 2, row, 1, 1, GridBagConstraints.LINE_END, padding,
+                    maxWeightx, weighty, GridBagConstraints.REMAINDER);
         }
 
         private void setupFieldForRemoteComponentsTab(final JTextField textField, String value) {
@@ -572,7 +578,8 @@ public class OptionsDialog extends AbstractWindow {
             }
 
             remoteServerData = remoteServerConfiguration.getServerList().get(0);
-            String finalInputValue = (textField instanceof JPasswordField) ? SecurityUtils.decodePasswordWithPrefix(value) : value;
+            String finalInputValue = (textField instanceof JPasswordField)
+                    ? SecurityUtils.decodePasswordWithPrefix(value) : value;
             textField.setText(finalInputValue);
             textField.getDocument().addDocumentListener(new DCDocumentListener() {
                 @Override
@@ -582,9 +589,9 @@ public class OptionsDialog extends AbstractWindow {
 
                     if (textField instanceof JPasswordField) {
                         remoteServerData.setPassword(textField.getText());
-                        _dcConfigurationUpdates.put(nodePath, SecurityUtils.encodePasswordWithPrefix(textField.getText()));
-                    }
-                    else {
+                        _dcConfigurationUpdates.put(nodePath,
+                                SecurityUtils.encodePasswordWithPrefix(textField.getText()));
+                    } else {
                         remoteServerData.setUsername(textField.getText());
                         _dcConfigurationUpdates.put(nodePath, remoteServerData.getUsername());
                     }
@@ -593,35 +600,11 @@ public class OptionsDialog extends AbstractWindow {
 
         }
 
-        private RemoteComponentServerType getRemoteComponentsServer() {
-            try {
-                InputStream inputStream = getDataCleanerConfigurationFileURI().openStream();
-                JaxbConfigurationReader jaxbConfigurationReader = new JaxbConfigurationReader();
-                Configuration configuration = jaxbConfigurationReader.unmarshall(inputStream);
-                List<Object> allProviders = configuration.getDescriptorProviders()
-                        .getCustomClassOrClasspathScannerOrRemoteComponents();
-
-                for (Object provider : allProviders) {
-                    if (provider instanceof RemoteComponentsType) {
-                        List<RemoteComponentServerType> servers = ((RemoteComponentsType) provider).getServer();
-
-                        // Only for the first server
-                        return servers == null || servers.isEmpty() ? null : servers.get(0);
-                    }
-                }
-            }
-            catch (IOException e) {
-                logger.warn(e.getMessage());
-            }
-
-            return null;
-        }
-
         private Component getDescriptionComponent() {
-            DCHtmlBox htmlBox = new DCHtmlBox("This dialog is for credentials setting of users registered at " +
-                    "<a href=\"http://datacleaner.org\">datacleaner.org</a>. <br><br>" +
-                    "Remote components are a cloud service providing new functions. " +
-                    "These remote components run at the server, consume provided input data and return the results. ");
+            DCHtmlBox htmlBox = new DCHtmlBox("This dialog is for credentials setting of users registered at "
+                    + "<a href=\"http://datacleaner.org\">datacleaner.org</a>. <br><br>"
+                    + "Remote components are a cloud service providing new functions. "
+                    + "These remote components run at the server, consume provided input data and return the results. ");
 
             return htmlBox;
         }

--- a/engine/core/src/main/java/org/datacleaner/configuration/RemoteServerConfigurationImpl.java
+++ b/engine/core/src/main/java/org/datacleaner/configuration/RemoteServerConfigurationImpl.java
@@ -26,8 +26,13 @@ import java.util.List;
  * Implementation of {@link RemoteServerConfiguration}.
  */
 public class RemoteServerConfigurationImpl implements RemoteServerConfiguration {
+
+    private final List<RemoteServerData> remoteServerDataList;
     private boolean showComponentsFromAllServers = false;
-    private List<RemoteServerData> remoteServerDataList = new ArrayList<>();
+
+    public RemoteServerConfigurationImpl() {
+        remoteServerDataList = new ArrayList<>();
+    }
 
     @Override
     public boolean isShowComponentsFromAllServers() {

--- a/engine/core/src/test/java/org/datacleaner/util/ws/ServiceSessionTest.java
+++ b/engine/core/src/test/java/org/datacleaner/util/ws/ServiceSessionTest.java
@@ -57,6 +57,8 @@ public class ServiceSessionTest extends TestCase {
         for (int i = 0; i < threads.length; i++) {
             threads[i].join();
         }
+        
+        session.close();
 
         assertEquals(threads.length, executionTimes.size());
 

